Control: function addition for the phase criterion

Spiros7bit opened the pull request:
<!-- Control: function addition for the phase condition. -->


#### Brief description of what is fixed or changed
I added a function to the TransferFunction class called phase_condition and has an input. 
This input can be a number that belongs to C, 
if it belongs to the Root Locus of the transfer function it returns True otherwise False. 

#### Other comments
* I think it's a good start for creating extensions over the Root Locus.

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* physics.control
    * Added a new function for phase condition.
<!-- END RELEASE NOTES -->

#### Reference
[Root Locus Analysis](https://en.wikipedia.org/wiki/Root_locus_analysis)

Spiros7bit committed:
Control: function addition for the phase criterion

I added a function to the TransferFunction class
called phase_criterion(z) and has an input.
This input can be a number that belongs to C and
if it belongs to the Root Locus of the transfer
function it returns True otherwise False. I think
it's a good start for creating extensions over the
Root Locus.

Example:

>>> system2 = TransferFunction((1), ((s**2+100*s+2600)*(s+25)*s), s)
>>> z1 = 37.1+65.2*I
>>> system2.phase_criterion(z1)
True

modified:   sympy/physics/control/lti.py
modified:   sympy/physics/control/tests/test_lti.py

moorepants commented:
Can you point to something that explains what the phase criteria is?

This may be a useful addition, but you'll need to make it work with fully symbolic transfer functions.

Spiros7bit commented:
Yes, it was a mistake not to analyze the phase criterion further. [Τhe sum of the angles from the open-loop zeros to the point minus the angles from the open-loop poles to the point has to be equal to π](https://en.wikipedia.org/wiki/Root_locus_analysis#Angle_condition). The point mentioned in the above wikipedia extension is the number we give to the function. If this criterion is true then this point is rootlocus' part. 
I also understand that the name I gave is incorrect. The correct name is "phase condition" (not "phase criterion").

Spiros7bit commented:
I will correct the name and the errors that were thrown out. And I'll commit again.

Spiros7bit committed:
Control: function addition for the phase condition

I added a function to the TransferFunction class
called phase_condition(z) and has an input.
This input can be a number that belongs to C and
if it belongs to the Root Locus of the transfer function
phase_condition returns True otherwise False.
I think it's a good start for creating extensions over the Root Locus.

Example:

>>> system2 = TransferFunction((1), ((s**2+100*s+2600)*(s+25)*s), s)
>>> z1 = 37.1+65.2*I
>>> system2.phase_condition(z1)
True

modified:   sympy/physics/control/lti.py
modified:   sympy/physics/control/tests/test_lti.py

Note:
This is the same commit with previous but with correct name (condition, not criterion)

Reference:
https://en.wikipedia.org/wiki/Root_locus_analysis

Spiros7bit committed:
Delete some whitespaces.

This code quality error occur:
AssertionError: File contains trailing whitespace: /home/runner/work/sympy/sympy/sympy/physics/control/lti.py

Spiros7bit committed:
Delete some whitespaces

Modified the code to pass the code-quality test

Spiros7bit committed:
author: add Spiros7bit to .mailmap

Spiros7bit commented:
> Can you point to something that explains what the phase criteria is?
> 
> This may be a useful addition, but you'll need to make it work with fully symbolic transfer functions.

When I started writing this function, I wanted to make it as symbolic as possible. When I say symbolic, I mean using only the sympy tools. Could you refer me or give me more information about what you are thinking about?
Thank you for your time.

moorepants commented:
You can start by having no numbers in your unit tests, only symbols.

Spiros7bit committed:
Function work with symbolic input numbers.

Notes: (I forgot to mention that the relation should be equal to the odd multiple of pi).
I modified the function to work with symbolic input numbers. Some constraints arose:
1. I was able to do the implementation for the first 8 odd multiples.
2. The processing of the symbolic characters needs a little more work.

Spiros7bit committed:
fixing an error

Substrate the greek letter π with pi.

Spiros7bit committed:
Fixing some erros from test.

There was an error about the parentheses in the test_phase_condition.

Spiros7bit commented:
> You can start by having no numbers in your unit tests, only symbols.

I modified the function to be able to process symbolic incomes. But it still needs some corrections.
